lib/browser.js

```diff
@@ -221,25 +196,67 @@ var Browser = function (id, fullName, /* capturedBrowsers */ collection, emitter
 
     this.lastResult.add(result)
 
-    emitter.emit('spec_complete', this, result)
-    refreshNoActivityTimeout()
+    this.emitter.emit('spec_complete', this, result)
+    this.refreshNoActivityTimeout()
   }
 
-  this.serialize = function () {
+  serialize () {
     return {
       id: this.id,
       name: this.name,
       isReady: this.state === READY
     }
   }
 
-  this.execute = function (config) {
-    activeSockets.forEach(function (socket) {
-      socket.emit('execute', config)
-    })
+  execute (config) {
+    this.activeSockets.forEach((socket) => socket.emit('execute', config))
 
     this.state = EXECUTING
-    refreshNoActivityTimeout()
+    this.refreshNoActivityTimeout()
+  }
+
+  getActiveSocketsIds () {
+    return this.activeSockets.map((s) => s.id).join(', ')
+  }
+
+  disconnect (reason) {
+    this.state = DISCONNECTED
+    this.disconnectsCount++
+    this.log.warn('Disconnected (%d times)' + (reason || ''), this.disconnectsCount)
+    this.emitter.emit('browser_error', this, 'Disconnected' + reason)
+    this.collection.remove(this)
+  }
+
+  refreshNoActivityTimeout () {
+    if (this.noActivityTimeout) {
+      this.clearNoActivityTimeout()
+
+      this.noActivityTimeoutId = this.timer.setTimeout(() => {
+        this.lastResult.totalTimeEnd()
+        this.lastResult.disconnected = true
+        this.disconnect(', because no message in ' + this.noActivityTimeout + ' ms.')
+        this.emitter.emit('browser_complete', this)
+      }, this.noActivityTimeout)
+    }
+  }
+
+  clearNoActivityTimeout () {
+    if (this.noActivityTimeout) {
+      if (this.noActivityTimeoutId) {
+        this.timer.clearTimeout(this.noActivityTimeoutId)
+        this.noActivityTimeoutId = null
+      }
+    }
+  }
+
+  bindSocketEvents (socket) {
+    // TODO: check which of these events are actually emitted by socket
```

Comment:
I think this change, while probably a good direction, should be separate so we don't get sources of errors mixed up.

Reply:
Methods in `class` declarations are not `enumerable` properties, so `bindAll` does not work over ES2015 classes... We'll either have to change `bindAll` implementation somehow or do this one. I think being explicit here is better (it took me some time to figure out why it was failing...). I can extract this into separate commit explaining the change if you prefer it?

Reply:
Yes, thanks, I do think making this change separately and before the classes would be good strategy.

Reply:
Extracted this change into a separate commit: https://github.com/karma-runner/karma/pull/2948/commits/a51505a57c0a85439d69df74be3651b32ab2e7e6.

Reply:
Thanks!  I think for a change of this magnitude and one that (necessarily) differs in style we should have other folks like @dignifiedquire weigh in.   The 'browser' changes look good to me.

Reply:
I am all for the explicit version 👍